tests/rest/client/test_relations.py

```diff
@@ -1070,7 +1033,143 @@ def assert_annotations(bundled_aggregations: JsonDict) -> None:
+        """
+        Bundled aggregations should get applied to the latest thread event.
+        """
+        self._send_relation(RelationTypes.THREAD, "m.room.test")
+        channel = self._send_relation(RelationTypes.THREAD, "m.room.test")
+        thread_2 = channel.json_body["event_id"]
+
+        self._send_relation(
+            RelationTypes.ANNOTATION, "m.reaction", "a", parent_id=thread_2
+        )
+
+        def assert_thread(bundled_aggregations: JsonDict) -> None:
+            self.assertEqual(2, bundled_aggregations.get("count"))
+            self.assertTrue(bundled_aggregations.get("current_user_participated"))
+            # The latest thread event has some fields that don't matter.
+            self.assert_dict(
+                {
+                    "content": {
+                        "m.relates_to": {
+                            "event_id": self.parent_id,
+                            "rel_type": RelationTypes.THREAD,
+                        }
+                    },
+                    "event_id": thread_2,
+                    "sender": self.user_id,
+                    "type": "m.room.test",
+                },
+                bundled_aggregations.get("latest_event"),
+            )
+            # Check the unsigned field on the latest event.
+            self.assert_dict(
+                {
+                    "m.relations": {
+                        RelationTypes.ANNOTATION: {
+                            "chunk": [
+                                {"type": "m.reaction", "key": "a", "count": 1},
+                            ]
+                        },
+                    }
+                },
+                bundled_aggregations["latest_event"].get("unsigned"),
+            )
+
+        self._test_bundled_aggregations(RelationTypes.THREAD, assert_thread, 10)
+
+    def test_thread_loop(self) -> None:
+        """Ensure that bogus events do not cause the bundled aggregations code to iterate forever."""
+        last_event = self.parent_id
+
+        # Disable the validation to pretend this came over federation.
+        with patch(
+            "synapse.handlers.message.EventCreationHandler._validate_event_relation",
+            new=lambda self, event: make_awaitable(None),
+        ):
+            for _ in range(2):
+                channel = self._send_relation(
+                    RelationTypes.THREAD, "m.room.test", parent_id=last_event
+                )
+                last_event = channel.json_body["event_id"]
+
+        # Fetch the thread root, to get the bundled aggregation for the thread.
+        relations = self._get_bundled_aggregations()
+
+        # The latest event should have bundled aggregations.
+        self.assertIn(RelationTypes.THREAD, relations)
+        thread_summary = relations[RelationTypes.THREAD]
+        self.assertIn("latest_event", thread_summary)
+
+        # The latest event should not have any bundled aggregations (since it
+        # only has a thread attached.
```

Comment:
```suggestion
        # The latest event should not have any bundled aggregations (since the
        # only relation to it is another thread, which is invalid).
```

likewise below.

Reply:
823a6b11c05dae9d2fde70045cb7e4e63b6766d4